File: src/test/kotlin/nl/marcenschede/zinder/matcher/business/MatcherDBInMemoryImpl.java
package nl.marcenschede.zinder.matcher.business;

import nl.marcenschede.zinder.matcher.business.MatcherDB;
import nl.marcenschede.zinder.matcher.business.Profile;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class MatcherDBInMemoryImpl implements MatcherDB {
    private Map<String, Profile> accountMap = new HashMap<>();

    @Override
    public void store(Profile profile) {
        accountMap.put(profile.getId(), profile);
    }

    @Override
    public Optional<Profile> findAccountById(String profileId) {
        return Optional.ofNullable(accountMap.get(profileId));
    }

    @Override
    public Collection<Profile> findMatchingAccounts(Profile requestProfile) {
        return accountMap.entrySet().stream()
                .map(Map.Entry::getValue)
                .filter(profile ->  profile.matches(requestProfile))
                .collect(Collectors.toSet());
    }
}

File: src/test/kotlin/nl/marcenschede/zinder/matcher/business/MatcherDBInMemoryImpl.kt
package nl.marcenschede.zinder.matcher.business

import java.util.*

fun Profile.matches(requestProfile: ProfileRequest): Boolean {
    val tagIsAvailableInProfile = { requestedTag :String ->
        val tagMatchesIgnoreCase = { tag : String -> tag.equals(requestedTag, ignoreCase = true) }

        tags.stream().anyMatch(tagMatchesIgnoreCase)
    }

    return requestProfile.tags.stream()
            .allMatch(tagIsAvailableInProfile)
}

class MatcherDBInMemoryImpl : MatcherDB {
    private val accountMap = HashMap<String, Profile>()

    override fun store(profile: Profile) {
        accountMap.put(profile.id, profile)
    }

    override fun findAccountById(profileId: String): Optional<Profile> {
        return Optional.ofNullable(accountMap[profileId])
    }

    override fun findMatchingAccounts(requestProfile: ProfileRequest): Collection<Profile> {
        return accountMap.values
                .filter { profile -> profile.matches(requestProfile) }
                .toList()
    }
}
